bash-cache.sh

```diff
@@ -278,31 +256,30 @@ bc::cache() {
     $_bc_enabled || { bc::orig::%func% "$@"; return; }
     ( bc::_cleanup & ) # Clean up stale caches in the background
 
-    local func="%func%" ttl="%ttl%" refresh="%refresh%" env=(%env%) args=("$@") cache_read_loc
+    local func="%func%" ttl="%ttl%" refresh="%refresh%" env=(%env%) args=("$@") exit cache_read_loc
     bc::_set_cache_read_loc
 
-    # Read from cache - capture output once to avoid races
-    # Note redirecting stderr to /dev/null comes first to suppress errors due to missing stdin
-    local out err exit
-    bc::_read_input out 2>/dev/null < "${cache_read_loc}/out" || true
-    bc::_read_input err 2>/dev/null < "${cache_read_loc}/err" || true
-    bc::_read_input exit 2>/dev/null < "${cache_read_loc}/exit" || true
-
-    if [[ -z "$exit" ]]; then
-      # No cache, execute in foreground
+    while true; do
+      { {
+        exit=$(< "${cache_read_loc}/exit")
+        # if exit is missing/empty we raced with a cleanup, disregard cache
+        if [[ -n "$exit" ]]; then
+          if (( refresh > 0 )) && ! bc::_newer_than "${cache_read_loc}/exit" "$refresh"; then
+            # Cache exists but is old, refresh in background
+            ( bc::_write_cache & )
+          fi
+          cat <&3 >&1 # stdout
+          cat <&4 >&5 # stderr
+          return "$exit"
+        fi
+      # This syntax preserves any existing file descriptors that might be open
+      # see https://mywiki.wooledge.org/FileDescriptor#Juggling_FDs
+      # Discard stderr first so missing /out / /err files don't generate noise
+      } 2>/dev/null 3<"${cache_read_loc}/out" 4<"${cache_read_loc}/err"; } 5>&2
```

Comment:
Well this is fucking wild.

Reply:
That is high praise 😛

I actually went and added a _third_ layer, so that all stderr in the inner block gets rerouted through fd5, rather than just the `/err` file contents (there shouldn't be any other stderr output, but I don't want to accidentally discard an unexpected error message).